libft/src/ft_getopt.c

```diff
@@ -0,0 +1,107 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   ft_getopt.c                                        :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: sgardner <stephenbgardner@gmail.com>       +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2018/04/03 16:20:06 by sgardner          #+#    #+#             */
+/*   Updated: 2018/04/10 07:09:04 by sgardner         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include <stdlib.h>
+#include <unistd.h>
+#include "ft_getopt.h"
+#include "libft.h"
+
+char		*g_optarg;
+int			g_optind = 1;
+int			g_opterr = 1;
+int			g_fderr = STDERR_FILENO;
+int			g_optopt;
+
+static int		opt_error(char *pname, const char *err, const char *optstring)
+{
+	char	*msg;
+	char	*tmp;
+	int		len;
+
+	if (g_opterr && *optstring != ':')
+	{
+		len = ft_strlen(pname) + ft_strlen(err) + 9;
+		if (!(msg = ft_memalloc(len)))
+			return ('?');
+		tmp = msg;
+		tmp = ft_stpcpy(tmp, pname);
+		tmp = ft_stpcpy(tmp, ": ");
+		tmp = ft_stpcpy(tmp, err);
+		tmp = ft_stpcpy(tmp, " -- ");
+		*tmp++ = g_optopt;
+		*tmp = '\n';
+		write(g_fderr, msg, len);
+		free(msg);
+	}
+	return ((*optstring == ':') ? ':' : '?');
+}
+
+static t_bool	set_arg(char *const av[], char **pos, const char *op)
+{
+	t_bool	optional;
+
+	g_optarg = NULL;
+	if (*(op + 1) != ':')
+		return (TRUE);
+	g_optarg = (**pos) ? *pos : av[g_optind];
+	optional = (*op && *(op + 2) == ':');
+	if (optional && g_optarg && *g_optarg == '-')
+		g_optarg = NULL;
+	else
+		++g_optind;
+	*pos = NULL;
+	return (g_optarg != NULL || optional);
+}
+
+static int		process_arg(char *const av[], const char *optstring)
+{
+	static char	*pos;
+	const char	*op;
+
+	if (!pos || !*pos)
+		pos = av[g_optind] + 1;
+	g_optopt = *pos++;
+	if (!*pos)
+		++g_optind;
+	if (!(op = ft_strchr(optstring, g_optopt)))
+		return (opt_error(av[0], "illegal option", optstring));
+	if (!set_arg(av, &pos, op))
+		return (opt_error(av[0], "option requires an argument", optstring));
+	return (g_optopt);
+}
+
+int				ft_getopt(int ac, char *const av[], const char *optstring)
+{
+	if (g_optind == 1)
+	{
+		g_optarg = NULL;
+		g_optopt = '\0';
+	}
+	while (g_optind < ac)
```

Comment:
Why while loop if there is always a return?

Reply:
It's a remnant from when it would look for flags in the entire array, before I settled on the return flag '\1'  GNU extension for optstrings beginning with '-'. It could just as easily be replaced with: 

`if (g_optind >= ac) { return (-1); }`

But it doesn't make much of a difference either way...